### ISSUE TITLE
Fail in repeated evaluation of 2nd derivative of a custom autograd function

### ISSUE THREAD

**Issue by jhrmnn** [pytorch/pytorch#18619]
## 🐛 Bug

Repeated evaluation of a 2nd derivative of a custom autograd function gives incorrect results.

## To Reproduce

```python
import torch


def mult1(x):
    return x.prod(dim=-1).prod(dim=-1)


class Mult(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = mult1(x)
        ctx.save_for_backward(x, y)
        return y

    @staticmethod
    def backward(ctx, grad_output):
        x, y = ctx.saved_tensors
        return (grad_output * y)[:, None, None] / x


mult2 = Mult.apply


def check_gradgrad_repeated(x, y):
    gy, = torch.autograd.grad(y[0], x, create_graph=True)
    ggy_1, = torch.autograd.grad(gy[0, 0, 0], x, retain_graph=True)
    gy, = torch.autograd.grad(y[0], x, create_graph=True)
    ggy_2, = torch.autograd.grad(gy[0, 0, 0], x, retain_graph=True)
    print(ggy_1[0, 0, 1], ggy_2[0, 0, 1])


x = torch.ones(2, 4, 4).requires_grad_()
check_gradgrad_repeated(x, mult1(x))
check_gradgrad_repeated(x, mult2(x))
```

The output is:

```
tensor(1.) tensor(1.)
tensor(1.) tensor(-0.)
```

## Expected behavior

The results should be the same on repeated evaluation.

## Environment

PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Debian GNU/Linux 9.8 (stretch)
GCC version: (Homebrew gcc 5.5.0_4) 5.5.0
CMake version: version 3.13.4

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: NVS 310

Nvidia driver version: 390.87
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.1
[pip] torch==1.0.1.post2
[conda] Could not collect

**Comment by jhrmnn** [comment]
Assuming this has been confirmed as a bug—before it gets fixed, any suggestion of a quick-fix workaround?

**Comment by ezyang** [comment]
I haven't tried running the test case but it seems pretty compelling. It might be related to the view created by `y[0]`, can you rewrite your code not to use it?

**Comment by jhrmnn** [comment]
I was able to simplify the minimal example to just a 2-vector, and without any indexing.

Also, on third evaluation of the first derivative, Pytorch aborts with an error message that doesn't make much sense:

```python
import torch


def mult1(x):
    return x.prod(dim=-1)


class Mult(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = mult1(x)
        ctx.save_for_backward(x, y)
        return y

    @staticmethod
    def backward(ctx, grad_output):
        x, y = ctx.saved_tensors
        return (grad_output * y) / x


mult2 = Mult.apply

print(
    'gradcheck',
    torch.autograd.gradcheck(mult2, torch.randn(2).double().requires_grad_()),
    torch.autograd.gradgradcheck(mult2, torch.randn(2).double().requires_grad_()),
)


def check_gradgrad_repeated(x, y):
    gy, = torch.autograd.grad(y, x, create_graph=True)
    ggy_1, = torch.autograd.grad(gy, x, torch.ones_like(gy), retain_graph=True)
    gy, = torch.autograd.grad(y, x, create_graph=True)
    ggy_2, = torch.autograd.grad(gy, x, torch.ones_like(gy), retain_graph=True)
    print(ggy_1, ggy_2)
    torch.autograd.grad(y, x, create_graph=True)


x = torch.tensor([1.0, 2.0]).requires_grad_()
check_gradgrad_repeated(x, mult1(x))
check_gradgrad_repeated(x, mult2(x))
```

Output:

```
gradcheck True True
tensor([1., 1.]) tensor([1., 1.])
tensor([1., 1.]) tensor([-2.0000, -0.5000])
Traceback (most recent call last):
  File "test.py", line 41, in <module>
    check_gradgrad_repeated(x, mult2(x))
  File "test.py", line 36, in check_gradgrad_repeated
    torch.autograd.grad(y, x, create_graph=True)
  File "/srv/hdd/jhermann/.cache/conda/envs/dl-qmc/lib/python3.7/site-packages/torch/autograd/__init__.py", line 145, in grad
    inputs, allow_unused)
RuntimeError: One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior.
```

I would guess that this has to do with how objects derived from `torch.autograd.Function` handle caching.

**Comment by jhrmnn** [comment]
I was able to track this down to the output of the forward pass being stored in `ctx.saved_tensors` for the backward pass. In many cases, reusing the output of a function is the fastest way to compute its gradient. Here's the minimal example:

```python
import torch


class Double(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = x ** 2
        ctx.save_for_backward(x, y)
        return y

    @staticmethod
    def backward(ctx, grad_output):
        x, _ = ctx.saved_tensors
        return grad_output * 2 * x


# this is equivalent, but uses the output of .forward() in .backward()
class Double2(Double):
    @staticmethod
    def backward(ctx, grad_output):
        x, y = ctx.saved_tensors
        return grad_output * 2 * y / x


double = Double.apply
double2 = Double2.apply

x = torch.tensor(2).double().requires_grad_()

print(
    'gradcheck',
    torch.autograd.gradcheck(double, x),
    torch.autograd.gradgradcheck(double, x),
    torch.autograd.gradcheck(double2, x),
    torch.autograd.gradgradcheck(double2, x),
)

y = double(x)
torch.autograd.grad(y, x, create_graph=True)
torch.autograd.grad(y, x)
print('double() ok')

y = double2(x)
torch.autograd.grad(y, x, create_graph=True)
torch.autograd.grad(y, x)
print('double2() ok')
```

Output:

```
gradcheck True True True True
double() ok
Traceback (most recent call last):
  File "test.py", line 44, in <module>
    torch.autograd.grad(y, x)
  File "/srv/hdd/jhermann/.cache/conda/envs/dl-qmc/lib/python3.7/site-packages/torch/autograd/__init__.py", line 145, in grad
    inputs, allow_unused)
  File "/srv/hdd/jhermann/.cache/conda/envs/dl-qmc/lib/python3.7/site-packages/torch/autograd/function.py", line 76, in apply
    return self._forward_cls.backward(self, *args)
  File "test.py", line 20, in backward
    x, y = ctx.saved_tensors
RuntimeError: Trying to backward through the graph a second time, but the buffers have already been freed. Specify retain_graph=True when calling backward the first time.
```

**Comment by jhrmnn** [comment]
Actually, storing any intermediate results from the forward pass seems to trigger the error:

```python
import torch


class Double(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = x ** 3
        ctx.save_for_backward(x, y)
        y = y + 1
        return y

    @staticmethod
    def backward(ctx, grad_output):
        x, _ = ctx.saved_tensors
        return grad_output * 3 * x ** 2


class Double2(Double):
    @staticmethod
    def backward(ctx, grad_output):
        x, y = ctx.saved_tensors
        assert torch.all(y / x == x ** 2)
        return grad_output * 3 * y / x


double = Double.apply
double2 = Double2.apply

x = torch.tensor(2).double().requires_grad_()

print('gradcheck double', torch.autograd.gradcheck(double, x))
print('gradgradcheck double', torch.autograd.gradgradcheck(double, x))
print('gradcheck double2', torch.autograd.gradcheck(double2, x))
print('gradgradcheck double2', torch.autograd.gradgradcheck(double2, x))
```

Output:

```
gradcheck double True
gradgradcheck double True
gradcheck double2 True
Traceback (most recent call last):
  File "test.py", line 34, in <module>
    print('gradgradcheck double2', torch.autograd.gradgradcheck(double2, x))
  File "/srv/hdd/jhermann/.cache/conda/envs/dl-qmc/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 302, in gradgradcheck
    return gradcheck(new_func, tupled_inputs + tupled_grad_outputs, eps, atol, rtol, raise_exception)
  File "/srv/hdd/jhermann/.cache/conda/envs/dl-qmc/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 205, in gradcheck
    'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
  File "/srv/hdd/jhermann/.cache/conda/envs/dl-qmc/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 185, in fail_test
    raise RuntimeError(msg)
RuntimeError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[-1.6448]], dtype=torch.float64)
analytical:tensor([[0.8224]], dtype=torch.float64)
```

**Comment by jhrmnn** [comment]
After some poking around, I understand now that this is a natural consequence of `forward()` not tracking any operations, so that if intermediate results are used in `backward()`, Pytorch cannot know the computational graph and perform the derivatives. The correct approach is to implement also higher derivatives manually. So this is not a bug, but perhaps insufficient documentation.

**Comment by ezyang** [comment]
Thanks; I haven't completely understood your example, but I certainly agree that if you produce intermediate results from forwards, you must specify appropriate backwards formulas for them as well (much in the same way that if you wrote a forward function with multiple outputs, of course you'd have to write backwards for each output.)

I agree there is definitely something to be improved in the documentation here. If you do come up with a doc patch, please CC me and I'll review it.

**Comment by apaszke** [comment]
This is certainly true in case of intermediate values (we might want to forbid adding those to `save_for_backward`, and require assigning them to `ctx` to make it clear that they're different), but you really _should_ be able to use the output of forward, which has been correctly saved, as part of your backward, so I don't think this is a problem limited to documentation only.

**Comment by jhrmnn** [comment]
@apaszke Am I right that currently there is no way to save the output of `forward` such that `backward` can correctly use it?

**Comment by apaszke** [comment]
`save_for_backward` exists exactly for that purpose, what you see is a very severe bug on our side.

**Comment by gchanan** [comment]
@ezyang can you take a look?

**Comment by ezyang** [comment]
I observed that this works:

```
import torch

class Double(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = x ** 2 
        ctx.save_for_backward(x, y)
        return y 

    @staticmethod
    def backward(ctx, grad_output):
        x, y = ctx.saved_tensors
        return grad_output * 2 * y / x 


double = Double.apply

x = torch.tensor(2).double().requires_grad_()

y = double(x)
g = torch.autograd.grad(y, x, create_graph=True)
torch.autograd.grad(y, x)
print('double() ok')
```

Somehow, `g` becoming dead causes the buffers for `y` to become dead.  I actually wonder if this is related to #22983 (I was testing on master; I should see what happens if I test with my copy.)

**Comment by ezyang** [comment]
There are a number of examples here, and they have varying levels of "should work" or "known problem". Quick summary:

1. https://github.com/pytorch/pytorch/issues/18619#issue-427080798 the original example should work, but doesn't.
2. https://github.com/pytorch/pytorch/issues/18619#issuecomment-479919095 works if you hold onto the grad before you run the second backwards. It should not be necessary to hold onto the grad in this way.
3. https://github.com/pytorch/pytorch/issues/18619#issuecomment-479929030 is not expected to work, if you save an intermediate computation, we don't expect you to be able to backpropagate through it. But maybe we can give a better error message in that case.

**Comment by ezyang** [comment]
I've confirmed that https://github.com/pytorch/pytorch/issues/18619#issuecomment-479919095 is fixed by #22983. I'll add a test for this in the PR.

**Comment by ezyang** [comment]
I've also confirmed https://github.com/pytorch/pytorch/issues/18619#issue-427080798 is fixed by #22983. Two for one package deal! So all that's left is to document saving intermediate computations.

**Comment by ezyang** [comment]
Fixed by https://github.com/pytorch/pytorch/pull/22983
